<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Clinseq Team">
  <link rel="shortcut icon" href="../../img/favicon.ico">
  <title>Alignment - Autoseq 2.0</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Alignment";
    var mkdocs_page_input_path = "lld/alignment.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../../js/jquery-2.1.1.min.js" defer></script>
  <script src="../../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/rust.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href="../.." class="icon icon-home"> Autoseq 2.0</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="../..">Autoseq</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../autoseq/">Quick Start</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../pipeline/">Work Flow</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../..">Autoseq 2.0</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../..">Docs</a> &raquo;</li>
    
      
    
    <li>Alignment</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h3 id="align_library">align_library</h3>
<p>Align fastq files for a PE library</p>
<ul>
<li>param remove_duplicates:</li>
<li>param pipeline:</li>
<li>param fq1_files:</li>
<li>param fq2_files:</li>
<li>param lib:</li>
<li>param ref:</li>
<li>param outdir:</li>
<li>param maxcores:</li>
<li>return:</li>
</ul>
<pre><code class="py">def align_library(pipeline, fq1_files, fq2_files, clinseq_barcode, ref, outdir, maxcores=1,
                  remove_duplicates=True):
    if not fq2_files:
        logging.debug(&quot;lib {} is SE&quot;.format(clinseq_barcode))
        return align_se(pipeline, fq1_files, clinseq_barcode, ref, outdir, maxcores, remove_duplicates)
    else:
        logging.debug(&quot;lib {} is PE&quot;.format(clinseq_barcode))
        return align_pe(pipeline, fq1_files, fq2_files, clinseq_barcode, ref, outdir, maxcores, remove_duplicates)
</code></pre>

<h3 id="align_se">align_se</h3>
<p>Align single end data</p>
<ul>
<li>param pipeline:</li>
<li>param fq1_files:</li>
<li>param lib:</li>
<li>param ref:</li>
<li>param outdir:</li>
<li>param maxcores:</li>
<li>param remove_duplicates:</li>
<li>return: bam </li>
</ul>
<pre><code class="py">def align_se(pipeline, fq1_files, clinseq_barcode, ref, outdir, maxcores, remove_duplicates=True):
    logging.debug(&quot;Aligning files: {}&quot;.format(fq1_files))
    fq1_abs = [normpath(x) for x in fq1_files]
    fq1_trimmed = []
    for fq1 in fq1_abs:
        skewer = Skewer()
        skewer.input1 = fq1
        skewer.input2 = None
        skewer.output1 = outdir + &quot;/skewer/{}&quot;.format(os.path.basename(fq1))
        skewer.output2 = outdir + &quot;/skewer/unused-dummyfq2-{}&quot;.format(os.path.basename(fq1))
        skewer.stats = outdir + &quot;/skewer/skewer-stats-{}.log&quot;.format(os.path.basename(fq1))
        skewer.threads = maxcores
        skewer.jobname = &quot;skewer/{}&quot;.format(os.path.basename(fq1))
        skewer.scratch = pipeline.scratch
        skewer.is_intermediate = True
        fq1_trimmed.append(skewer.output1)
        pipeline.add(skewer)

    cat1 = Cat()
    cat1.input = fq1_trimmed
    cat1.output = outdir + &quot;/skewer/{}_1.fastq.gz&quot;.format(clinseq_barcode)
    cat1.jobname = &quot;cat/{}&quot;.format(clinseq_barcode)
    cat1.is_intermediate = False
    pipeline.add(cat1)

    bwa = Bwa()
    bwa.input_fastq1 = cat1.output
    bwa.input_reference_sequence = ref
    bwa.remove_duplicates = remove_duplicates

    library_id = parse_prep_id(clinseq_barcode)
    sample_string = compose_sample_str(extract_unique_capture(clinseq_barcode))

    bwa.readgroup = &quot;\&quot;@RG\\tID:{rg_id}\\tSM:{rg_sm}\\tLB:{rg_lb}\\tPL:ILLUMINA\&quot;&quot;.format(\
        rg_id=clinseq_barcode, rg_sm=sample_string, rg_lb=library_id)

    bwa.threads = maxcores
    bwa.output = &quot;{}/{}.bam&quot;.format(outdir, clinseq_barcode)
    bwa.scratch = pipeline.scratch
    bwa.jobname = &quot;bwa/{}&quot;.format(clinseq_barcode)
    bwa.is_intermediate = False
    pipeline.add(bwa)

    return bwa.output
</code></pre>

<h3 id="align_pe">align_pe</h3>
<p>align paired end data</p>
<ul>
<li>param pipeline:</li>
<li>param fq1_files:</li>
<li>param fq2_files:</li>
<li>param lib:</li>
<li>param ref:</li>
<li>param outdir:</li>
<li>param maxcores:</li>
<li>param remove_duplicates:</li>
<li>return:</li>
</ul>
<pre><code class="py">def align_pe(pipeline, fq1_files, fq2_files, clinseq_barcode, ref, outdir, maxcores=1, remove_duplicates=True):
    fq1_abs = [normpath(x) for x in fq1_files]
    fq2_abs = [normpath(x) for x in fq2_files]
    logging.debug(&quot;Trimming {} and {}&quot;.format(fq1_abs, fq2_abs))
    pairs = [(fq1_abs[k], fq2_abs[k]) for k in range(len(fq1_abs))]

    fq1_trimmed = []
    fq2_trimmed = []

    for fq1, fq2 in pairs:
        skewer = Skewer()
        skewer.input1 = fq1
        skewer.input2 = fq2
        skewer.output1 = outdir + &quot;/skewer/libs/{}&quot;.format(os.path.basename(fq1))
        skewer.output2 = outdir + &quot;/skewer/libs/{}&quot;.format(os.path.basename(fq2))
        skewer.stats = outdir + &quot;/skewer/libs/skewer-stats-{}.log&quot;.format(os.path.basename(fq1))
        skewer.threads = maxcores
        skewer.jobname = &quot;skewer/{}&quot;.format(os.path.basename(fq1))
        skewer.scratch = pipeline.scratch
        skewer.is_intermediate = True
        fq1_trimmed.append(skewer.output1)
        fq2_trimmed.append(skewer.output2)
        pipeline.add(skewer)

    cat1 = Cat()
    cat1.input = fq1_trimmed
    cat1.output = outdir + &quot;/skewer/{}-concatenated_1.fastq.gz&quot;.format(clinseq_barcode)
    cat1.jobname = &quot;cat1/{}&quot;.format(clinseq_barcode)
    cat1.is_intermediate = True
    pipeline.add(cat1)

    cat2 = Cat()
    cat2.input = fq2_trimmed
    cat2.jobname = &quot;cat2/{}&quot;.format(clinseq_barcode)
    cat2.output = outdir + &quot;/skewer/{}-concatenated_2.fastq.gz&quot;.format(clinseq_barcode)
    cat2.is_intermediate = True
    pipeline.add(cat2)

    bwa = Bwa()
    bwa.input_fastq1 = cat1.output
    bwa.input_fastq2 = cat2.output
    bwa.input_reference_sequence = ref
    bwa.remove_duplicates = remove_duplicates

    library_id = parse_prep_id(clinseq_barcode)
    sample_string = compose_sample_str(extract_unique_capture(clinseq_barcode))

    bwa.readgroup = &quot;\&quot;@RG\\tID:{rg_id}\\tSM:{rg_sm}\\tLB:{rg_lb}\\tPL:ILLUMINA\&quot;&quot;.format(\
        rg_id=clinseq_barcode, rg_sm=sample_string, rg_lb=library_id)

    bwa.threads = maxcores
    bwa.output = &quot;{}/{}.bam&quot;.format(outdir, clinseq_barcode)
    bwa.jobname = &quot;bwa/{}&quot;.format(clinseq_barcode)
    bwa.scratch = pipeline.scratch
    bwa.is_intermediate = False
    pipeline.add(bwa)

    return bwa.output
</code></pre>

<h3 id="fq_trimming">fq_trimming</h3>
<ul>
<li>return concatinated fastq files</li>
</ul>
<pre><code class="py">def fq_trimming(pipeline, fq1_files, fq2_files, clinseq_barcode, ref, outdir, maxcores=1):
    fq1_abs = [normpath(x) for x in fq1_files]
    fq2_abs = [normpath(x) for x in fq2_files]
    logging.debug(&quot;Trimming {} and {}&quot;.format(fq1_abs, fq2_abs))
    pairs = [(fq1_abs[k], fq2_abs[k]) for k in range(len(fq1_abs))]

    fq1_trimmed = []
    fq2_trimmed = []

    for fq1, fq2 in pairs:
        skewer = Skewer()
        skewer.input1 = fq1
        skewer.input2 = fq2
        skewer.output1 = outdir + &quot;/skewer/libs/{}&quot;.format(os.path.basename(fq1))
        skewer.output2 = outdir + &quot;/skewer/libs/{}&quot;.format(os.path.basename(fq2))
        skewer.stats = outdir + &quot;/skewer/libs/skewer-stats-{}.log&quot;.format(os.path.basename(fq1))
        skewer.threads = maxcores
        skewer.jobname = &quot;skewer/{}&quot;.format(os.path.basename(fq1))
        skewer.scratch = pipeline.scratch
        skewer.is_intermediate = True
        fq1_trimmed.append(skewer.output1)
        fq2_trimmed.append(skewer.output2)
        pipeline.add(skewer)

    cat1 = Cat()
    cat1.input = fq1_trimmed
    cat1.output = outdir + &quot;/skewer/{}-concatenated_1.fastq.gz&quot;.format(clinseq_barcode)
    cat1.jobname = &quot;cat1/{}&quot;.format(clinseq_barcode)
    cat1.is_intermediate = True
    pipeline.add(cat1)

    cat2 = Cat()
    cat2.input = fq2_trimmed
    cat2.jobname = &quot;cat2/{}&quot;.format(clinseq_barcode)
    cat2.output = outdir + &quot;/skewer/{}-concatenated_2.fastq.gz&quot;.format(clinseq_barcode)
    cat2.is_intermediate = True
    pipeline.add(cat2)

    return cat1.output, cat2.output
</code></pre>

<h3 id="bwa">Bwa</h3>
<pre><code class="py">class Bwa(Job):
    def __init__(self):
        Job.__init__(self)
        self.input_fastq1 = None  # input ports must start with &quot;input&quot;
        self.input_fastq2 = None
        self.input_reference_sequence = None
        self.mark_secondary = None
        self.remove_duplicates = True
        self.readgroup = None
        self.output = None  # output ports must start with &quot;output&quot;, can be &quot;output_metrics&quot;, &quot;output&quot;, etc
        self.duplication_metrics = None
        self.jobname = &quot;bwa&quot;

    def command(self):
        bwalog = self.output + &quot;.bwa.log&quot;
        samblasterlog = self.output + &quot;.samblaster.log&quot;
        tmpprefix = &quot;{}/{}&quot;.format(self.scratch, uuid.uuid4())

        return &quot;bwa mem -M -v 1 &quot; + \
               required(&quot;-R &quot;, self.readgroup) + \
               optional(&quot;-t &quot;, self.threads) + \
               required(&quot; &quot;, self.input_reference_sequence) + \
               required(&quot; &quot;, self.input_fastq1) + \
               optional(&quot;&quot;, self.input_fastq2) + \
               required(&quot;2&gt;&quot;, bwalog) + \
               &quot;| samblaster -M --addMateTags &quot; + \
               conditional(self.remove_duplicates, &quot;--removeDups&quot;) + \
               optional(&quot;--metricsFile &quot;, self.duplication_metrics) + \
               required(&quot;2&gt;&quot;, samblasterlog) + \
               &quot;| samtools view -Sb -u - &quot; + \
               &quot;| samtools sort &quot; + \
               required(&quot;-T &quot;, tmpprefix) + \
               optional(&quot;-@ &quot;, self.threads) + \
               required(&quot;-o &quot;, self.output) + \
               &quot; - &quot; + \
               &quot; &amp;&amp; samtools index &quot; + self.output + \
               &quot; &amp;&amp; cat {} {}&quot;.format(bwalog, samblasterlog) + \
               &quot; &amp;&amp; rm {} {}&quot;.format(bwalog, samblasterlog)
</code></pre>

<h3 id="skewer">Skewer</h3>
<pre><code class="py">class Skewer(Job):
    def __init__(self):
        Job.__init__(self)
        self.input1 = None
        self.input2 = None
        self.output1 = None
        self.output2 = None
        self.stats = None
        self.jobname = &quot;skewer&quot;

    def command(self):
        if not self.output1.endswith(&quot;.gz&quot;) or not self.output2.endswith(&quot;.gz&quot;):
            raise ValueError(&quot;Output files need to end with .gz&quot;)

        tmpdir = os.path.join(self.scratch, &quot;skewer-&quot; + str(uuid.uuid4()))
        prefix = &quot;{}/skewer&quot;.format(tmpdir)

        if self.input2:
            out_fq1 = prefix + &quot;-trimmed-pair1.fastq.gz&quot;
            out_fq2 = prefix + &quot;-trimmed-pair2.fastq.gz&quot;
        else:
            out_fq1 = prefix + &quot;-trimmed.fastq.gz&quot;
            out_fq2 = &quot;&quot;

        out_stats = prefix + &quot;-trimmed.log&quot;

        mkdir_cmd = &quot;mkdir -p {}&quot;.format(tmpdir)

        skewer_cmd = &quot;skewer -z &quot; + \
                     optional(&quot;-t &quot;, self.threads) + &quot; --quiet &quot; + \
                     required(&quot;-o &quot;, prefix) + \
                     required(&quot;&quot;, self.input1) + \
                     optional(&quot;&quot;, self.input2)
        copy_output_cmd = &quot;cp &quot; + out_fq1 + &quot; &quot; + self.output1 + \
            conditional(self.input2, &quot; &amp;&amp; cp &quot; + out_fq2 + &quot; &quot; + self.output2)

        copy_stats_cmd = &quot;cp &quot; + out_stats + &quot; &quot; + self.stats
        rm_cmd = &quot;rm -r {}&quot;.format(tmpdir)
        return &quot; &amp;&amp; &quot;.join([mkdir_cmd, skewer_cmd, copy_output_cmd, copy_stats_cmd, rm_cmd])
</code></pre>
              
            </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/ClinSeq/autoseq" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
      
    </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme.js" defer></script>
      <script src="../../search/main.js" defer></script>

</body>
</html>
