<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Clinseq Team">
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Work Flow - Autoseq 2.0</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Work Flow";
    var mkdocs_page_input_path = "pipeline.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/rust.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Autoseq 2.0</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Autoseq</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../autoseq/">Quick Start</a>
	    </li>
          
            <li class="toctree-l1 current">
		
    <a class="current" href="./">Work Flow</a>
    <ul class="subnav">
            
    <li class="toctree-l2"><a href="#liqbio">LiqBio</a></li>
    

    <li class="toctree-l2"><a href="#alascca">Alascca</a></li>
    

    </ul>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Autoseq 2.0</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Work Flow</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="liqbio">LiqBio</h1>
<p><em>Non UMI pipeline</em></p>
<ol>
<li><strong>Preprocessing</strong><ul>
<li>Trimming and Filtering<ul>
<li><a href="https://github.com/relipmoc/skewer">skewer</a></li>
</ul>
</li>
<li>Mapping to Reference<ul>
<li><a href="http://bio-bwa.sourceforge.net/bwa.shtml">BWA</a></li>
</ul>
</li>
<li>Indel Realignment<ul>
<li><a href="https://gatkforums.broadinstitute.org/gatk/discussion/7156/howto-perform-local-realignment-around-indels">GATK4</a></li>
</ul>
</li>
<li>Merge Bams and Markdups<ul>
<li><a href="https://broadinstitute.github.io/picard/">Picard</a></li>
</ul>
</li>
</ul>
</li>
<li><strong>Germline Variant Calling</strong><ul>
<li>SNV and small INDEL<ul>
<li><a href="https://software.broadinstitute.org/gatk/documentation/tooldocs/3.8-0/org_broadinstitute_gatk_tools_walkers_haplotypecaller_HaplotypeCaller.php">GATK4-HaplotypeCaller</a></li>
<li><a href="https://github.com/Illumina/strelka">Strelka Germline</a></li>
</ul>
</li>
<li>Combining Variants<ul>
<li><a href="https://software.broadinstitute.org/gatk/documentation/tooldocs/3.8-0/org_broadinstitute_gatk_tools_walkers_variantutils_CombineVariants.php">GATK3-CombineVariants</a></li>
</ul>
</li>
</ul>
</li>
<li><strong>Somatic Variant Calling</strong><ul>
<li>SNV and small INDEL<ul>
<li><a href="https://github.com/AstraZeneca-NGS/VarDictJava">VarDict</a></li>
<li><a href="https://github.com/Illumina/strelka">Strelka Somatic</a></li>
<li><a href="https://gatkforums.broadinstitute.org/gatk/discussion/11136/how-to-call-somatic-mutations-using-gatk4-mutect2">GATK4-Mutect2</a></li>
<li><a href="http://varscan.sourceforge.net/">Varscan</a></li>
</ul>
</li>
<li>Combine and Filter Variants(Ensembl Approach)<ul>
<li><a href="https://github.com/bioinform/somaticseq">SomaticSeq</a></li>
</ul>
</li>
<li>Annatotaion<ul>
<li><a href="https://www.ensembl.org/info/docs/tools/vep/index.html">VeP-95</a></li>
</ul>
</li>
</ul>
</li>
<li><strong>Structural Variant Calling</strong><ul>
<li>SV - INS, INV, DEL, TRA, DUP<ul>
<li><a href="https://github.com/tomwhi/svcaller">svcaller - In-house developed</a></li>
<li><a href="https://github.com/vpc-ccg/svict">svict</a></li>
<li><a href="https://github.com/arq5x/lumpy-sv">Lumpy</a></li>
<li><a href="https://github.com/walaj/svaba">svaba</a></li>
</ul>
</li>
</ul>
</li>
<li><strong>Copy Number Variation</strong><ul>
<li><a href="https://cnvkit.readthedocs.io/en/stable/index.html">CNVkit</a></li>
<li><a href="https://bioconductor.org/packages/devel/bioc/html/PureCN.html">PureCN  in-house developed script</a></li>
<li>liqbio_CNA - for visualization</li>
</ul>
</li>
<li><strong>QC Metrics</strong><ul>
<li>Fastq - Quality Check<ul>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></li>
</ul>
</li>
<li>Coverage<ul>
<li>alascca coverage hist</li>
<li>coverage qc call</li>
</ul>
</li>
<li>hsmetrics<ul>
<li><a href="https://broadinstitute.github.io/picard/">Picard</a></li>
</ul>
</li>
<li>OXOG metrics<ul>
<li><a href="https://broadinstitute.github.io/picard/">Picard</a></li>
</ul>
</li>
<li>Insert size <ul>
<li><a href="https://broadinstitute.github.io/picard/">Picard</a></li>
</ul>
</li>
<li>Sequencing Depth <ul>
<li>sambamba-depth</li>
</ul>
</li>
<li>MSI<ul>
<li><a href="https://github.com/ding-lab/msisensor">msisensor</a></li>
<li><a href="https://bitbucket.org/uwlabmed/msings/src/master/">mSINGS</a> - Tumor only</li>
</ul>
</li>
<li>Contamination Test<ul>
<li>GATK conEst</li>
</ul>
</li>
<li>MultiQC</li>
</ul>
</li>
</ol>
<p><em>UMI pipeline</em></p>
<ol>
<li>UMI processing<ul>
<li>Trimming and Filtering<ul>
<li><a href="https://github.com/relipmoc/skewer">skewer</a></li>
</ul>
</li>
<li>Fastq to Bam<ul>
<li><a href="http://fulcrumgenomics.github.io/fgbio/tools/latest/FastqToBam.html">fgbio-FastqToBam</a></li>
</ul>
</li>
<li>Alignment of Unmapped Bam - 1<ul>
<li><a href="http://bio-bwa.sourceforge.net/bwa.shtml">BWA</a></li>
</ul>
</li>
<li>Realignment - 1<ul>
<li><a href="https://gatkforums.broadinstitute.org/gatk/discussion/7156/howto-perform-local-realignment-around-indels">GATK4</a></li>
</ul>
</li>
<li>Group reads by umi<ul>
<li><a href="http://fulcrumgenomics.github.io/fgbio/tools/latest/GroupReadsByUmi.html">fgbio-GroupReadsByUmi</a></li>
</ul>
</li>
<li>Call Duplex Consensus Reads<ul>
<li><a href="http://fulcrumgenomics.github.io/fgbio/tools/latest/CallDuplexConsensusReads.html">CallDuplexConsensusReads</a></li>
</ul>
</li>
<li>Alignment - 2<ul>
<li><a href="http://bio-bwa.sourceforge.net/bwa.shtml">BWA</a></li>
</ul>
</li>
<li>Realignment - 2<ul>
<li><a href="https://gatkforums.broadinstitute.org/gatk/discussion/7156/howto-perform-local-realignment-around-indels">GATK4</a></li>
</ul>
</li>
<li>Filter Consensus Reads<ul>
<li><a href="http://fulcrumgenomics.github.io/fgbio/tools/latest/FilterConsensusReads.html">fgbio-FilterConsensusReads</a></li>
</ul>
</li>
<li>Clip Overlapping Reads<ul>
<li><a href="http://fulcrumgenomics.github.io/fgbio/tools/latest/ClipBam.html">fgbio-ClipBam</a></li>
</ul>
</li>
<li>Mark Duplicates (bam file from Realignment-1)<ul>
<li><a href="https://broadinstitute.github.io/picard/">Picard</a></li>
</ul>
</li>
</ul>
</li>
<li>Germline Variant Calling (bam file from clip overlapping bam)<ul>
<li>SNV and small INDEL<ul>
<li><a href="https://software.broadinstitute.org/gatk/documentation/tooldocs/3.8-0/org_broadinstitute_gatk_tools_walkers_haplotypecaller_HaplotypeCaller.php">GATK4-HaplotypeCaller</a></li>
<li><a href="https://github.com/Illumina/strelka">Strelka Germline</a></li>
</ul>
</li>
<li>Combining Variants<ul>
<li><a href="https://software.broadinstitute.org/gatk/documentation/tooldocs/3.8-0/org_broadinstitute_gatk_tools_walkers_variantutils_CombineVariants.php">GATK3-CombineVariants</a></li>
</ul>
</li>
</ul>
</li>
<li>Somatic Variant Calling (bam file from clip overlapping bam)<ul>
<li>SNV and small INDEL<ul>
<li><a href="https://github.com/AstraZeneca-NGS/VarDictJava">VarDict</a></li>
<li><a href="https://github.com/Illumina/strelka">Strelka Somatic</a></li>
<li><a href="https://gatkforums.broadinstitute.org/gatk/discussion/11136/how-to-call-somatic-mutations-using-gatk4-mutect2">GATK4-Mutect2</a></li>
<li><a href="http://varscan.sourceforge.net/">Varscan</a></li>
</ul>
</li>
<li>Combine and Filter Variants(Ensembl Approach)<ul>
<li><a href="https://github.com/bioinform/somaticseq">SomaticSeq</a></li>
</ul>
</li>
<li>Annatotaion<ul>
<li><a href="https://www.ensembl.org/info/docs/tools/vep/index.html">VeP-95</a></li>
</ul>
</li>
</ul>
</li>
<li>Structural Variant Calling <ul>
<li>SV - INS, INV, DEL, TRA, DUP<ul>
<li><a href="https://github.com/tomwhi/svcaller">svcaller - In-house variant caller based on delly Algorythm</a></li>
<li><a href="https://github.com/vpc-ccg/svict">svict</a></li>
<li><a href="https://github.com/arq5x/lumpy-sv">Lumpy</a></li>
<li><a href="https://github.com/walaj/svaba">svaba</a></li>
</ul>
</li>
</ul>
</li>
<li>Copy Number Variation<ul>
<li><a href="https://cnvkit.readthedocs.io/en/stable/index.html">CNVkit</a></li>
<li><a href="https://bioconductor.org/packages/devel/bioc/html/PureCN.html">PureCN  in-house developed script</a></li>
<li>liqbio_CNA - for visualization</li>
</ul>
</li>
<li>QC Metrics<ul>
<li>Fastq - Quality Check<ul>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></li>
</ul>
</li>
<li>Coverage<ul>
<li>alascca coverage hist</li>
<li>coverage qc call</li>
</ul>
</li>
<li>hsmetrics<ul>
<li><a href="https://broadinstitute.github.io/picard/">Picard</a></li>
</ul>
</li>
<li>OXOG metrics<ul>
<li><a href="https://broadinstitute.github.io/picard/">Picard</a></li>
</ul>
</li>
<li>Insert size <ul>
<li><a href="https://broadinstitute.github.io/picard/">Picard</a></li>
</ul>
</li>
<li>Sequencing Depth <ul>
<li>sambamba-depth</li>
</ul>
</li>
<li>MSI<ul>
<li><a href="https://github.com/ding-lab/msisensor">msisensor</a></li>
<li><a href="https://bitbucket.org/uwlabmed/msings/src/master/">mSINGS</a> - Tumor only</li>
</ul>
</li>
<li>Contamination Test<ul>
<li>GATK conEst</li>
</ul>
</li>
<li>MultiQC</li>
</ul>
</li>
</ol>
<h1 id="alascca">Alascca</h1>
<p><em>Alascca pipeline needs to be fixed as per new changes done in clinseq pipeline</em></p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="../autoseq/" class="btn btn-neutral" title="Quick Start"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/ClinSeq/autoseq" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../autoseq/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
