<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Clinseq Team">
  <link rel="shortcut icon" href="../../img/favicon.ico">
  <title>Qc - Autoseq 2.0</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Qc";
    var mkdocs_page_input_path = "lld/qc.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../../js/jquery-2.1.1.min.js" defer></script>
  <script src="../../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/python.min.js"></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/rust.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href="../.." class="icon icon-home"> Autoseq 2.0</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="../..">Autoseq</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../autoseq/">Quick Start</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../pipeline/">Work Flow</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../..">Autoseq 2.0</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../..">Docs</a> &raquo;</li>
    
      
    
    <li>Qc</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="microsatelite-instability">Microsatelite Instability</h1>
<hr />
<h3 id="msisensor">MsiSensor</h3>
<pre><code class="py">class MsiSensor(Job):
    def __init__(self):
        Job.__init__(self)
        self.msi_sites = None
        self.input_normal_bam = None
        self.input_tumor_bam = None
        self.output = None
        self.jobname = &quot;msisensor&quot;

    def command(self):
        output_prefix = &quot;{scratch}/msisensor-{uuid}&quot;.format(scratch=self.scratch,
                                                            uuid=uuid.uuid4())
        output_table = &quot;{}&quot;.format(output_prefix)
        output_dis = &quot;{}_dis&quot;.format(output_prefix)
        output_germline = &quot;{}_germline&quot;.format(output_prefix)
        output_somatic = &quot;{}_somatic&quot;.format(output_prefix)

        return &quot;msisensor msi &quot; + \
               required(&quot;-d &quot;, self.msi_sites) + \
               required(&quot;-n &quot;, self.input_normal_bam) + \
               required(&quot;-t &quot;, self.input_tumor_bam) + \
               required(&quot;-o &quot;, output_prefix) + \
               required(&quot;-b &quot;, self.threads) + \
               &quot; &amp;&amp; cp {} {}&quot;.format(output_prefix, self.output) + \
               &quot; &amp;&amp; rm {} {} {} {}&quot;.format(output_table, output_dis,
                                           output_germline, output_somatic)
</code></pre>

<h3 id="msisensorscan">MsiSensorScan</h3>
<pre><code class="py">class MsiSensorScan(Job):
    def __init__(self):
        Job.__init__(self)
        self.input_fasta = None
        self.output = None
        self.homopolymers_only = True
        self.jobname = &quot;msisensor-scan&quot;

    def command(self):
        return &quot;msisensor scan &quot; + \
               required(&quot;-d &quot;, self.input_fasta) + \
               required(&quot;-o &quot;, self.output) + \
               conditional(self.homopolymers_only, &quot; -p 1 &quot;)

</code></pre>

<h3 id="intersectmsisites">IntersectMsiSites</h3>
<pre><code class="py">class IntersectMsiSites(Job):
    def __init__(self):
        Job.__init__(self)
        self.input_msi_sites = None
        self.target_bed = None
        self.output_msi_sites = True
        self.jobname = &quot;msi-intersect&quot;

    def command(self):
        return &quot;intersect-msi-sites.sh &quot; + \
               required(&quot; &quot;, self.input_msi_sites) + \
               required(&quot; &quot;, self.target_bed) + \
               required(&quot; &quot;, self.output_msi_sites)
</code></pre>

<h3 id="msings">Msings</h3>
<pre><code class="py">class Msings(Job):
    def __init__(self):
        Job.__init__(self)
        self.input_fasta = None
        self.outdir = None
        self.output = None
        self.msings_baseline = None
        self.msings_bed = None
        self.msings_intervals = None
        self.input_bam = None

        self.jobname = &quot;msings&quot;

    def command(self):
        return &quot;run_msings.sh&quot; + \
               required(&quot;-b &quot;, self.msings_bed) + \
               required(&quot;-f &quot;, self.input_fasta) + \
               required(&quot;-i &quot;, self.msings_intervals) + \
               required(&quot;-n &quot;, self.msings_baseline) + \
               required(&quot;-o &quot;, self.outdir) + \
               required(&quot; &quot;, self.input_bam)
</code></pre>

<h1 id="purecn">PureCN</h1>
<hr />
<p>PureCN  is a purity and ploidy aware copy number caller for cancer samples inspired by
the ABSOLUTE algorithm</p>
<p>Determining output here purely as a means of determining when the job has
completed; PureCN itself determines the output file paths based on the
specified "--output" and "--sampleid" arguments.</p>
<pre><code class="py">class PureCN(Job):
    def __init__(self, input_seg=None, input_vcf=None, tumorid=None, outdir=None, gcgene_file=None,
                 minpurity=0.05, hzdev=0.1, seg_sdev=None, genome=&quot;hg19&quot;, postopt=False):
        Job.__init__(self)
        self.input_seg = input_seg
        self.input_vcf = input_vcf
        self.tumorid = tumorid  # should be the same as the ID in the seg file
        self.outdir = outdir
        self.gcgene_file = gcgene_file
        self.minpurity = minpurity
        self.hzdev = hzdev
        self.seg_sdev = seg_sdev
        self.genome = genome
        self.postopt = postopt

        self.output = &quot;{}/{}_genes.csv&quot;.format(
            self.outdir, self.tumorid)

        self.jobname = &quot;purecn&quot;

    def command(self):

        # activating conda env
        activate_cmd = &quot;source activate purecn-env&quot;

        # running PureCN
        running_cmd = &quot;PureCN.R &quot; + required(&quot;--out &quot;, self.outdir) + \
                       required(&quot;--sampleid &quot;, self.tumorid) + \
                       required(&quot;--segfile &quot;, self.input_seg) + \
                       required(&quot;--vcf &quot;, self.input_vcf) + \
                       required(&quot;--gcgene &quot;, self.gcgene_file) + \
                       required(&quot;--genome &quot;, self.genome) + \
                       optional(&quot;--minpurity &quot;, self.minpurity) + \
                       optional(&quot;--hzdev &quot;, self.hzdev) + \
                       optional(&quot;--segfilesdev &quot;, self.seg_sdev) + \
                       conditional(self.postopt, &quot;--postoptimize&quot;)

        # deactivating the conda env
        deactivate_cmd = &quot;source deactivate&quot;

        return &quot; &amp;&amp; &quot;.join([activate_cmd, running_cmd, deactivate_cmd])
</code></pre>

<h1 id="quality-check">Quality Check</h1>
<hr />
<h3 id="heterzygoteconcordance">HeterzygoteConcordance</h3>
<p>Heterzygote Concoradance module customly added to GATKv3.5 by Klevebring. Jar file is located in autoseq-scripts github repo.</p>
<pre><code class="py">class HeterzygoteConcordance(Job):
    def __init__(self):
        Job.__init__(self)
        self.input_vcf = None
        self.input_bam = None
        self.reference_sequence = None
        self.target_regions = None
        self.normalid = None
        self.filter_reads_with_N_cigar = True
        self.output = None
        self.jobname = &quot;hzconcordance&quot;

    def command(self):
        return &quot;gatk-klevebring -T HeterozygoteConcordance &quot; + \
               required(&quot;-R &quot;, self.reference_sequence) + \
               required(&quot;-V &quot;, self.input_vcf) + \
               required(&quot;-I &quot;, self.input_bam) + \
               required(&quot;-sid &quot;, self.normalid) + \
               optional(&quot;-L &quot;, self.target_regions) + \
               conditional(self.filter_reads_with_N_cigar, &quot;--filter_reads_with_N_cigar&quot;) + \
               required(&quot;-o &quot;, self.output)
</code></pre>

<h3 id="fastqc">FastQC</h3>
<pre><code class="py">class FastQC(Job):
    def __init__(self):
        Job.__init__(self)
        self.input = None
        self.output = None
        self.outdir = None
        self.extract = False
        self.jobname = &quot;fastqc&quot;

    def command(self):
        return &quot;fastqc -o {} &quot;.format(self.outdir) + \
               conditional(self.extract, &quot;--extract&quot;) + \
               &quot; --nogroup {}&quot;.format(self.input)
</code></pre>

<h3 id="multiqc">MultiQC</h3>
<pre><code class="py">class MultiQC(Job):
    def __init__(self):
        Job.__init__(self)
        self.input_files = None
        self.search_dir = None
        self.output = None
        self.report_title = None
        self.data_format = 'json'
        self.jobname = &quot;multiqc&quot;

    def command(self):
        required(&quot;input_files&quot;, self.input_files)
        required(&quot;output_base&quot;, self.output)
        required(&quot;dir_to_search&quot;, self.search_dir)

        basefn = os.path.basename(self.output)
        odir = os.path.dirname(self.output)
        return &quot;multiqc &quot; + \
               required(&quot;&quot;, self.search_dir) + \
               required(&quot;-o &quot;, odir) + \
               optional(&quot;-n &quot;, basefn) + \
               optional(&quot;-k &quot;, self.data_format) + \
               optional(&quot;-i &quot;, self.report_title) + \
               &quot; --data-dir --zip-data-dir -v -f&quot;
</code></pre>

<h3 id="sambambadepth">SambambaDepth</h3>
<pre><code class="py">class SambambaDepth(Job):
    def __init__(self):
        Job.__init__(self)
        self.input = None
        self.targets_bed = None
        self.coverage_thresholds = [30, 50, 70, 100, 200, 300]
        self.output = None
        self.jobname = &quot;sambamba-depth&quot;

    def command(self):
        return &quot;sambamba depth region&quot; + \
               repeat(&quot;-T &quot;, self.coverage_thresholds) + \
               required(&quot;--regions &quot;, self.targets_bed) + \
               required(&quot;&quot;, self.input) + \
               required(&quot;&gt;&quot;, self.output)
</code></pre>

<h3 id="bedtoolscoveragehistogram">BedtoolsCoverageHistogram</h3>
<pre><code class="py">class BedtoolsCoverageHistogram(Job):
    def __init__(self):
        Job.__init__(self)
        self.input_bam = None
        self.input_bed = None
        self.tag = None
        self.output = None

    def command(self):
        tag_cmd = &quot;&quot;
        if self.tag:
            tag_cmd = &quot;echo \&quot;# {}\&quot; &gt;&gt; {} \n&quot;.format(self.tag, self.output)
        return &quot;echo \&quot;# bedtools-coverage-hist: {}\&quot;&quot;.format(self.input_bam) + \
               required(&quot;d&gt;&quot;, self.output) + &quot;\n&quot; + \
               tag_cmd + \
               &quot;bedtools coverage -hist &quot; + \
               required(&quot;-a &quot;, self.input_bed) + \
               required(&quot;-b &quot;, self.input_bam) + \
               &quot;|grep \&quot;^all\&quot; &quot; + required(&quot;&gt;&gt; &quot;, self.output)
</code></pre>

<h3 id="coveragehistogram">CoverageHistogram</h3>
<pre><code class="py">class CoverageHistogram(Job):
    def __init__(self):
        Job.__init__(self)
        self.input_bam = None
        self.input_bed = None
        self.min_basequal = None
        self.output = None

    def command(self):
        return &quot;target_coverage_histogram.py &quot; + \
               required(&quot;--targets &quot;, self.input_bed) + \
               required(&quot; &quot;, self.input_bam) + \
               optional(&quot;--min_basequal &quot;, self.min_basequal) + \
               required(&quot;&gt; &quot;, self.output)
</code></pre>

<h3 id="coveragecaveat">CoverageCaveat</h3>
<pre><code class="py">class CoverageCaveat(Job):
    def __init__(self):
        Job.__init__(self)
        self.input_histogram = None
        self.output = None
        self.high_thresh_fraction = 0.95
        self.high_thresh_fold_cov = 100
        self.low_thresh_fraction = 0.95
        self.low_thresh_fold_cov = 50

    def command(self):
        return &quot;extract_coverage_caveat.py &quot; + \
               required(&quot; &quot;, self.input_histogram) + \
               required(&quot;--high-thresh-fraction &quot;, self.high_thresh_fraction) + \
               required(&quot;--high-thresh-fold-cov &quot;, self.high_thresh_fold_cov) + \
               required(&quot;--low-thresh-fraction &quot;, self.low_thresh_fraction) + \
               required(&quot;--low-thresh-fold-cov &quot;, self.low_thresh_fold_cov) + \
               required(&quot;&gt; &quot;, self.output)
</code></pre>
              
            </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/ClinSeq/autoseq" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
      
    </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme.js" defer></script>
      <script src="../../search/main.js" defer></script>

</body>
</html>
